anndata/core/anndata.py

```diff
@@ -571,23 +575,29 @@ def _init_as_view(self, adata_ref: 'AnnData', oidx: Index, vidx: Index):
                 'Currently, you cannot index repeatedly into a backed AnnData, '
                 'that is, you cannot make a view of a view.')
         self._isview = True
+        if isinstance(oidx, (int, np.integer)): oidx = slice(oidx, oidx+1, 1)
+        if isinstance(vidx, (int, np.integer)): vidx = slice(vidx, vidx+1, 1)
```

Comment:
This touches on #145, since it effectively does what I said I didn't want and maintains the dimensionality of all constituent objects when the container object is subset. For example: `adata[1, :].layers["counts"]` is now a two dimensional ndarray.

This did make it much easier to get everything working, since I know which axis of all the contained array aligns to which axis of the AnnData object without having to check either's dimensions.

However, this makes some tests fail, since there are effectively checks on the dimensionality of `X`. To get around this I modified the accessor for `X` to return a one dimensional array if any dimension of `X` had a length of 1. This has opened a new can of worms with inconsistencies between `X` and `layers` as well as `adata[0, :].X.shape == adata[[0], :].X.shape` (which previously was not the case).

I'm trying to figure out what the right way forward here is. It makes our lives easier if contained arrays match up with the shape of their `AnnData`, but there's code that relies on this not being the case (`sc.pl.umap` for example, though this currently doesn't work with sparse layers anyways).